Yoast/YoastSEO.js
# Implement topic density assessment

## Summary

This PR can be summarized in the following changelog entry:

* Implements a topic density assessment that considers density of not only the keyword but also synonyms of the keyword, supplied by the user.

## Relevant technical choices:

* The keywordDensityAssessment is retained as it was.
* The keywordDensityAssessment is currently triggered all the time, but as soon as the Synonym Supply interface is ready it will only be triggered if no synonyms are supplied.
* The topicDensityAssessment has somewhat elevated boundaries compared to keywordDensityAssessment (by 0.5%), which is meant to allow those users who have already adjusted their writing style to keywordDensityAssessment to still get a green bullet with the new match functionality.

## Test instructions

This PR can be tested by following these steps:

* No changes have been implemented in WP-seo-premium yet, so please use browserified example.
* Paste a text and add a keyword. The keywordDensityAssessment should return meaningful result. The topicDensityAssessment should not be triggered.
* Add another word to the keyphrase, separated by a comma.
* The keywordDensityAssessment should change its result, most likely to 0% (unless you are extremely lucky and there is an exact combination of your keyword and synonym in your text).
* The topicDensityAssessment should appear and deliver a result similar to the original keywordDensityAssessment, but with somewhat higher %, in case both supplied words in the keyphrase occur in the text.
* Try adding multi-word items to the list, items with non-comma punctuation marks between words, words in different cases.

Fixes #1544
